framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java

```diff
@@ -516,12 +519,94 @@
     protected List<@FullyQualifiedName String> upstreamCheckerNames;
 
+    /**
+     * Tries to unwrap ProcessingEnvironment from proxy in IntelliJ >=2020.3.
+     *
+     * @param envProxy is dynamic proxy wrapping processing environment
+     * @return unwrapped processing environment, null if not successful
+     */
+    private static @Nullable ProcessingEnvironment unwrapIntelliJ(Object envProxy) {
+        InvocationHandler handler = Proxy.getInvocationHandler(envProxy);
+        try {
+            Field field = handler.getClass().getDeclaredField("val$delegateTo");
+            field.setAccessible(true);
+            Object o = field.get(handler);
+            if (o instanceof ProcessingEnvironment) {
+                return (ProcessingEnvironment) o;
+            }
+            return null;
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Gradle incremental processing unwrapping inspired by project Lombok.
+     *
+     * @param delegateClass is class in which we try to find delegate field
+     * @param env is processing environment wrapper
+     * @return unwrapped processing environment, null if not successful
+     */
+    private static @Nullable ProcessingEnvironment unwrapGradle(
+            Class<?> delegateClass, Object env) {
+        try {
+            Field field = delegateClass.getDeclaredField("delegate");
+            field.setAccessible(true);
+            Object o = field.get(env);
+            if (o instanceof ProcessingEnvironment) {
+                return (ProcessingEnvironment) o;
+            }
+            return null;
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            return null;
+        }
+    }
+
+    /**
+     * IntelliJ wraps processing environment in dynamic proxy to check for modifications done by
+     * annotation processors. But lots of functionality both here and in AbstractTypeProcessor call
+     * methods from javac that require processing environment to be
+     * com.sun.tools.javac.processing.JavacProcessingEnvironment and fail on proxy - thus in case
+     * proxy is used, we have to unwrap it.
+     *
+     * @param env is processing environment supplied to checker
+     * @return unwrapped environment if it is dynamic proxy, created by IntelliJ; original value in
+     *     all other cases
+     */
+    private static ProcessingEnvironment unwrapProcessingEnvironment(ProcessingEnvironment env) {
+        // equality, not instanceof corresponds to test in Trees and JavacTask
+        // comparison instead of equals is required by some tests in CheckerFramework CI
+        if (env.getClass().getName()
+                == "com.sun.tools.javac.processing.JavacProcessingEnvironment") {
+            return env;
+        }
+        // IntelliJ >2020.3 wraps processing environment in dynamic proxy...
+        if (Proxy.isProxyClass(env.getClass())) {
```

Comment:
I suggest putting this test within `unwrapIntelliJ`, so that all logic related to IntelliJ proxies is together.

Reply:
Moved